### ISSUE TITLE
IDEA: Possibly add "Unlikelihood Training" - the simplified version of DPO?

### ISSUE THREAD

**Issue by jukofyork** [tdrussell/qlora-pipe#57]
Not sure how hard it would be or where to start, but looking at the recent DPO changes, would it be possible to add "Unlikelihood Training":

https://arxiv.org/abs/1908.04319

It's basically DPO but with no reward model (ie: the reward ratios all set to 1 which simplifies to just the difference of the gradients).

The code looks to already allow alternate sequences to have the negative of the gradient added into the sun via some modulo-2 test, but not sure how you would set the rest up (possibly just allow a negative weighted dataset?).

It would be really useful for the extension of @tdrussell's "Instruct-Storywriter" method and save all sorts of other hacks I've been trying regarding zeroing the gradients of the special tokens, and basically allow the "prose signal" to not be so dominated by the "weird formatting" signal.

**Comment by tdrussell** [comment]
Any DPO-like training method should be pretty easy to add, especially if it doesn't need reference logits. The way the code works now, is that if the training example contains features prefixed with "rejected_", e.g. "rejected_input_ids", "rejected_attention_mask", etc, it forms batches where the first half the batch is the accepted examples, and the second half is rejected. Then at the point you compute the loss, it's easy to extract the accepted and rejected logits. For vanilla DPO, there's a special pipeline parallel schedule that additionally runs an extra forward pass, disabling the LoRA and setting some attributes so the loss layer knows to store the reference logits that it will then need on the second (normal) forward pass.

For unlikelihood training specifically, I'm not very familiar with it, or what benefits it might have over other techniques. If you just want to avoid needing the reference logits, I know there is a so-called "reference-free DPO", and also techniques like ORPO.

**Comment by jukofyork** [comment]
> For unlikelihood training specifically, I'm not very familiar with it, or what benefits it might have over other techniques. If you just want to avoid needing the reference logits, I know there is a so-called "reference-free DPO", and also techniques like ORPO.

Not sure if it will work, but the biggest problem I had with the "Instruct-Storywriter" method was the "formatting signal" was much stronger than the "prose signal", and however much data I had or however much I tried to mask the gradients of the special tokens; I couldn't reliably get a good model from it.

The idea of "unlikelihood training" is just to have a second set of "unwanted" data that will have the negated gradients added to the "wanted" data's loss, with the net effect (hopefully) being that if both are using formatting that is equally unusual; the "formatting signal" will cancel out and you should (hopefully) end up having a much clearer "wanted prose" vs "unwanted prose" signal left behind to follow.

I think with (very) careful regularisation it might be possible to do something similar just by training two LoRAs and then subtracting their multiplied-out differences, but gradient ascent in unbounded and has a tendency to "shoot off" into bad weight-space, and training both sides simultaneously might make it much easier to train.
